Modules/Applications/AppDomainTransform/app/otbDomainTransform.cxx

```diff
@@ -0,0 +1,453 @@
+
+        TInputImagePointer inImage = GetParameterImage("in");
+
+        inImage->UpdateOutputInformation();
+
+        // typedef TComplexImage for InverseFFTImageFilter input
+        // This a image type of std::complex<TInputPixel>
+        typedef otb::Image<
+          std::complex<TInputPixel>, 2 > TComplexImage;
+        //typedef TOutputImage for InverseFFTImageFilter output
+        typedef otb::Image< TOutputPixel >  TOutputImage;
+
+        // a unary functor to convert vectorimage to complex image
+        typedef itk::UnaryFunctorImageFilter
+          <TInputImage,
+           TComplexImage,
+           ToComplexPixel
+           <TInputImage::PixelType,
+            TComplexImage::PixelType> > UnaryFunctorImageFilter;
+
+        UnaryFunctorImageFilter::Pointer
+          unaryFunctorImageFilter = UnaryFunctorImageFilter::New();
+
+        if( shift)
+          {
+          typedef itk::FFTShiftImageFilter<
+            TInputImage,
+            TInputImage > FFTShiftFilterType;
+
+          FFTShiftFilterType::Pointer
+            fftShiftFilter = FFTShiftFilterType::New();
+          fftShiftFilter->SetInput( inImage );
+
+          fftShiftFilter->Update();
+
+          unaryFunctorImageFilter->SetInput(fftShiftFilter->GetOutput() );
+          }
+        else
+          {
+          unaryFunctorImageFilter->SetInput(inImage);
+          }
+
+        unaryFunctorImageFilter->Update();
+
+        //typedef itk::::InverseFFTImageFilter over TComplexImage
+        typedef itk::InverseFFTImageFilter
+          < TComplexImage,
+            TOutputImage > FFTFilter;
+        FFTFilter::Pointer invFilter = FFTFilter::New();
+        invFilter->SetInput( unaryFunctorImageFilter->GetOutput() );
+        invFilter->Update();
+
+        //set output image
+        SetParameterOutputImage<TOutputImage>("out", invFilter->GetOutput());
+        }
+      }
+    }
+
+  template<otb::Wavelet::Wavelet TWaveletOperator>
+  void DoWaveletTransform(const int dir,
+                          const unsigned int nlevels,
+                          const std::string inkey = "in",
+                          const std::string outkey = "out")
+    {
+    typedef otb::Image< TInputPixel >  TInputImage;
+    typedef otb::Image< TOutputPixel >  TOutputImage;
+    typedef typename TInputImage::Pointer TInputImagePointer;
+
+    TInputImagePointer inImage = GetParameterImage<TInputImage>(inkey);
+
+    inImage->UpdateOutputInformation();
+
+    if( dir == 0)
+      {
+      typedef otb::WaveletImageFilter<
+        TInputImage,
+        TOutputImage,
+        TWaveletOperator> TWaveletImageFilter;
+
+      typedef typename
+        TWaveletImageFilter::Pointer
+        TWaveletImageFilterPointer;
+      TWaveletImageFilterPointer waveletImageFilter =
+        TWaveletImageFilter::New();
+
+      waveletImageFilter->SetInput(inImage);
+      waveletImageFilter->SetNumberOfDecompositions(nlevels);
+      waveletImageFilter->Update();
```

Comment:
These filters are not streamed, but a better practice is to : don't call the update here, create an attribute (itk::ProcessObject::Pointer) to store waveletImageFilter after the end of DoExecute(), let the ApplicationEngine call Update() on the output image.